Proposals/20200511-HVPA-design.md

```diff
@@ -0,0 +1,262 @@
+
+**Example** - Consider a situation where each of the 2 replicas of an application running with 5GB memory has a base memory usage of 2GB because of caching, and the VPA memory recommendations are 6GB. Here the total recommended memory is 12GB. If, according to linearly divided bucket, HVPA needs to scale the applications to 3 replicas, then each replica would be just 4GB so that the total memory allocated to the application is set to 12GB. The memory available after deducting the base usage of all replicas would then be 6GB which is same as before the scaling.
+
+*Mitigation* - HVPA would scale the application to 3 replicas of size 6GB each. The total allocated memory would be 18GB, erring on side of stability rather than the cost.
+
+As a general rule of thumb, HVPA will set the size of each replica to the max of the actual VPA recommendation and the desired resources calculated as per bucket strategy. It is assumed by HVPA controller that the VPA is tuned properly as mentioned [here](####vpa-tuning).
+
+#### HPA domination
+HPA is a more macroscopic autoscaler than VPA, because HPA recommeds change in replicas while VPA recommends change in resources. Because of this, total CPU/memory resources recommended by HPA might tend to dominate under increasing load.
+
+*Mitigation* - Either do not use HPA by setting `.spec.hpa.deploy` to `false` , or set the threshold percentages for HPA accordingly, eg - above 100%. Another alternative is to rely on VPA tuning as explained [here](####vpa-tuning).
+
+## Design Details
+
+A user can provide an array of buckets described above in the [proposal](##proposal).
+For example, typical `scaleIntervals` would look like:
+```yaml
+spec:
+  scaleIntervals:
+  - maxCpu: "1",
+    maxMemory: "2G",
+    maxReplicas: "1",
+  - maxCpu: "20",
+    maxMemory: "30G",
+    maxReplicas: "5",
+```
+
+### Minimum required fields
+All the fields in a bucket - `maxCpu`, `maxMemory` and `maxReplicas` - are required to define a valid bucket. These fields are required to determine upper bound of the aggregate resources for that bucket.
+This upper bound of a bucket is considered as lower bound for the next bucket unless it is the first bucket in which case the lower bound is calculated using VPA's `minAllowed`s or `0`, and HPA's `minReplicas` or `1`.
+
+For subsequent buckets:
+* the minimum replicas is equal to one more than `maxReplicas` of previous bucket
+* minimum resource values are calculated such that aggregated maximum resources for previous bucket equals aggregated minimum resources for the bucket
+
+For the second bucket in the spec shown in above example, the minimum replicas is `2`, the minimum cpu is `500m` (`1 replica * 1 core / 2 replicas`) and the minimum memory is `1G` (`1 replica * 2G memory / 2 replicas`).
+
+### kube-apiserver scaling example:
+For kube-apiserver, a typical spec could look like:
+
+```yaml
+spec:
+  scaleIntervals:
+  - maxCpu: "500m",
```

Comment:
At least the first bucket needs some min values, I believe.

Reply:
Mentioned under section `Minimum required fields`:
```
...upper bound of a bucket is considered as lower bound for the next bucket unless it is the first bucket in which case the lower bound is calculated using VPA's `minAllowed`s or `0`, and HPA's `minReplicas` or `1`.
```